### ISSUE TITLE
Level 3: Approaching sub() method

### ISSUE THREAD

**Issue by jeanyeeee** [nus-cs2030/2122-s2#288]
# Summary
The question states:
For subtraction, say` m.sub(n)`, it is observed that it is equivalent to the addition, i.e. `(-n).add(m)`. That is to say, we need to negate the argument n (without invalidating it), and then add m to it. Use the` n `function in` AbstractNum`.

How to approach this? Making the number negative will invalidate it...should I create alternative methods for the `succ()` and `add()` which do not check if the arguments are valid?

Any help is appreciated tq!

**Comment by szeying02** [comment]
Making a negative AbstractNum won't invalid it! What I did was place the result into a new AbstractNum after following the question.

**Comment by Geraldworks** [comment]
isValid() method will always return true unless opt = Optional.<T>empty(). Hence, you can manually include an optional that holds a negative number into the AbstractNum superclass. Keep using succ() on this negative number until you reach Num.zero(). The terminating condition would thus be [trackerNum.equals(Num.zero())].

**Comment by hengyjj** [comment]
> # Summary
> The question states: For subtraction, say` m.sub(n)`, it is observed that it is equivalent to the addition, i.e. `(-n).add(m)`. That is to say, we need to negate the argument n (without invalidating it), and then add m to it. Use the`n`function in` AbstractNum`.
> 
> How to approach this? Making the number negative will invalidate it...should I create alternative methods for the `succ()` and `add()` which do not check if the arguments are valid?
> 
> Any help is appreciated tq!

Yes, you can convert them to a negative (-m) using `AbstractNum.n` method. What i did was I used `super.opt.map(AbstractNum.n)` and i will pass the outcome into the Num constructor to create a new Num. 
After getting your answer, you have to check if your result is valid or not by using the same method again using `output.opt.filter(AbstractNum.valid)`. Otherwise, you would return a `new Num(Optional.<Integer> empty())`. Hope this helps and atb!

**Comment by junbin-tan** [comment]
Hi @jeanyeeee: 

What I did for the `.sub(Num other)` is:
1) convert the number I want to subtract `other` to a negative number  --> name it as `result`
2) using a while loop I loop from 0 to this current Number 
3) within the while loop I use AbstractNum.s to add 1 to my result 

Idea behind it you loop keep adding +1 to the negative number eventually you will get the balance.

e.g. 3 - 2 
-2 + 1 = -1
-1 + 1 = 0
0 + 1 = 1

Hope it helps!

**Comment by jeanyeeee** [comment]
Hi @junbin-tan @hengyjj @szeying02 @Geraldworks 
Thank you for the pseudo-code, today i learned how to count and subtract HAHAHA
Turns out I had an `@Override isValid()` method which was invalidating my negative numbers...
After removing this method, 

I did @junbin-tan and @hengyjj method:
(For anyone that is still stuck)
1. have a `counter` initialized to zero and `result = Num(num.opt.map(AbstractNum.n))`
2. check if both arguments are valid
3. have a while loop (see above thread) to increment both the `counter` and `result` until the `counter` is `equal` to this
4. `return new Num(num.opt.filter(AbstractNum.valid))` or empty


Thank you sm, I passed this level :)
